src/calculator_definitions.jl

```diff
@@ -0,0 +1,247 @@
+AtomsCalculators.@generate_complement function AtomsCalculators.forces(system, calculator::Main.MyType; kwargs...)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition
+    return zeros(AtomsCalculators.default_force_eltype, length(system)) 
+end
+```
+
+Generate `forces` call from `forces!` definition
+
+```julia
+AtomsCalculators.@generate_complement function AtomsCalculators.forces!(f::AbstractVector, system, calculator::Main.MyOtherType; kwargs...)
+    @assert length(f) == length(system)
+    # we can ignore kwargs... or use them to tune the calculation
+    # or give extra information like pairlist
+
+    # add your own definition
+    for i in eachindex(f)
+        f[i] = zero(AtomsCalculators.default_force_eltype)
+    end
+
+    return f
+end
+```
+"""
+macro generate_complement(expr)
+    oldname = nothing
+    try
+        # determine either "force" or "force!"
+        # or fail if not present at all
+        oldname = "$(expr.args[1].args[1])"
+    catch _
+        error("Not valid input")
+    end
+
+    try
+        # check that has kwargs... support
+        has_kwargs = any( [ Symbol("...") == x.head  for x in expr.args[1].args[2].args ] )
+        !has_kwargs && error()
+    catch _
+        error("Call does not catch kwargs...")
+    end
+
+    calc_type = nothing
+    try
+        # expr.args[1].args[4] for "forces"
+        # and expr.args[1].args[5] for "forces!"
+        # is calculator based on definition.
+        # we can leave it to be the end now.
+        # But this needs to change, if we allow extra arguments.
+        calc_type = expr.args[1].args[end].args[2]
+    catch _
+        throw(error("Calculator does not have defined type"))
+    end
+
+    if oldname[end] == '!'
+        # generate "forces"
+        length(expr.args[1].args) != 5 && error("Number of inputs does not match the call")
+        name = oldname[begin:end-1] 
+        q = Meta.parse(
+            "function $name(system, calculator::$calc_type; kwargs...)
+                final_data = zeros( default_force_eltype, length(system) )
+                $oldname(final_data, system, calculator; kwargs...)
+                return final_data
+            end"
+        )  
+    else
+        # generate "forces!"
+        length(expr.args[1].args) != 4 && error("Number of inputs does not match the call")
+        name = oldname * "!"
+        q = Meta.parse(
+            "function $name(final_data::AbstractVector, system, calculator::$calc_type; kwargs...)
+                @assert length(final_data) == length(system)
+                final_data .= $oldname(system, calculator; kwargs...)
+                return final_data
+            end"
+        )
+    end
+    return quote
+        $expr
+        $q
+    end
+end
+
+
+## test functions to test interface
+
+
+"""
+    test_forces(sys, calculator; force_eltype::AbstractVector=default_force_eltype, kwargs...)
```

Comment:
this stuff should probably be defined in the test directory, no?

Reply:
per Zoom discussion, revising suggestion to just moving these to a separate file for now :)